src/python/pants/backend/scala/lint/scalafmt/rules.py

```diff
@@ -0,0 +1,175 @@
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import FieldSet, Target
+from pants.engine.unions import UnionRule
+from pants.jvm.jdk_rules import JdkSetup
+from pants.jvm.resolve.coursier_fetch import MaterializedClasspath, MaterializedClasspathRequest
+from pants.jvm.resolve.jvm_tool import JvmToolLockfileRequest, JvmToolLockfileSentinel
+from pants.testutil.rule_runner import logging
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+
+@dataclass(frozen=True)
+class ScalafmtFieldSet(FieldSet):
+    required_fields = (ScalaSourceField,)
+
+    source: ScalaSourceField
+
+    @classmethod
+    def opt_out(cls, tgt: Target) -> bool:
+        return tgt.get(SkipScalafmtField).value
+
+
+class ScalafmtRequest(ScalaFmtRequest, LintRequest):
+    field_set_type = ScalafmtFieldSet
+
+
+class ScalafmtToolLockfileSentinel(JvmToolLockfileSentinel):
+    options_scope = ScalafmtSubsystem.options_scope
+
+
+@dataclass(frozen=True)
+class SetupRequest:
+    request: ScalafmtRequest
+    check_only: bool
+
+
+@dataclass(frozen=True)
+class Setup:
+    process: Process
+    original_digest: Digest
+
+
+@logging
+@rule(level=LogLevel.DEBUG)
+async def setup_google_java_format(
+    setup_request: SetupRequest,
+    tool: ScalafmtSubsystem,
+    jdk_setup: JdkSetup,
+    bash: BashBinary,
+) -> Setup:
+    source_files, tool_classpath, conf_digest = await MultiGet(
+        Get(
+            SourceFiles,
+            SourceFilesRequest(field_set.source for field_set in setup_request.request.field_sets),
+        ),
+        Get(
+            MaterializedClasspath,
+            MaterializedClasspathRequest(
+                prefix="__toolcp",
+                lockfiles=(tool.resolved_lockfile(),),
+            ),
+        ),
+        # TODO: Figure out how to scan for the correct configuration file to use while also dealing with
+        #  the seeming requirement that `version` must be set in the configuration file.
+        Get(
+            Digest,
+            CreateDigest(
+                [
+                    FileContent(
+                        path=".scalafmt.conf",
+                        content=textwrap.dedent(
+                            f"""\
+                version = "{tool.version}"
+                runner.dialect = scala213
+                """
+                        ).encode(),
+                    )
+                ]
+            ),
+        ),
+    )
+
+    source_files_snapshot = (
+        source_files.snapshot
+        if setup_request.request.prior_formatter_result is None
+        else setup_request.request.prior_formatter_result
+    )
+
+    input_digest = await Get(
+        Digest,
+        MergeDigests(
+            [source_files_snapshot.digest, conf_digest, tool_classpath.digest, jdk_setup.digest]
+        ),
+    )
+
+    def add_check_args() -> Iterable[str]:
+        if setup_request.check_only:
+            yield "--list"
```

Comment:
Seems simpler to set after line 122 `if setup_request.check_only: args.append("--list")`. More direct